### DISCUSSION TITLE
How to unit test JS in TS based project

### DISCUSSION THREAD

**Discussion by unbreakabl3** [vmware/build-tools-for-vmware-aria#252] [Q&A] [answered]
Hello dear all,
I did try to play / test all available options to find a way how to unit test .js function in TS project. No success.

Description: I have a standard TS based project. There I have some `function.js`, which I want to unit test. Without exporting it, it's not possible to test. If I am exporting it, it's not converted properly to the normal action element in vRO with all the inputs, etc.
So, how to do that? :)

Thanks in advance.

**Comment by VenelinBakalov** [comment]
Hi @unbreakabl3 , for normal JS actions project you need to follow specific structure and put your tests under the src/test but I am not sure if the case is the same for TS oriented projects with JS actions
![image](https://github.com/vmware/build-tools-for-vmware-aria/assets/22296648/fc2574d6-dbea-4cf7-90cc-93c26d6c580d)

We might need some help from @Michaelpalacce

**Reply by unbreakabl3** [reply]
Hi, @VenelinBakalov .
Yeah. I did follow this folders structure and it's working, but only for JS based project. My question was how to do that for TS based project :) . I can guess its not possible, right?

**Comment by VenelinBakalov** [comment]
Yeap, I got the question, I was just suggesting in case you haven't tried that option. tbh I don't know the answer myself because I haven't had that case, in the meantime I will check back with the team to try and get some feedback. I will write here as soon as I have some answer from them :)

**Reply by unbreakabl3** [reply]
Appreciate

**Comment by Michaelpalacce** [comment]
@unbreakabl3 you can import them with System.getModules as you would normally in Orchestrator and perform the test

**Comment by unbreakabl3** [comment]
Hi, @Michaelpalacce .
I did it. It doesn't work in TS project. The import of System.getModules is testable in JS project only.
There is a chance I am doing something wrong, but I did try all possible scenarios I can think about :). 

If you do have a working example you can share, I will be very appreciative.

**Reply by Michaelpalacce** [reply]
Sorry for the late reply:

Test Code:

`sumAction.test.ts`
```typescript
const sumAction = System.getModule("com.vmware.pscoe.onboarding.sgenov.example").sumAction;

describe("Sum Action test", () => {
	it("should add two numbers", () => {
		expect(sumAction(1, 5)).toBe(6);
	});
});
```

Code to test: 

`sumAction.ts`
```typescript
( function ( numberOne: number, numberTwo: number ):number {
    return numberOne + numberTwo;
});
```
My code structure:
<img width="366" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/17056014/174a1c79-89ec-4b29-b464-fbc3f7e7be69">

my pom xml artifact data

<img width="453" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/17056014/0e9c2601-57dc-4c8e-8a04-f80f424e7a5d">

Output:

<img width="751" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/17056014/a9d92dc4-bc6d-4965-88ae-4ca3ed53aadd">

Feel free to mark this as an answer if it helped!

**Reply by unbreakabl3** [reply]
Hi, @Michaelpalacce.
Thank a lot for that example. But the described original problem is to test **JS** file in TS project. In my question, I want to test `function.js`. Not `function.ts`. In your example, you're testing TS file in TS project :)

**Reply by unbreakabl3** [reply]
If I will follow your example, but with testing JS file, so:


<img width="755" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/13556249/7ecf1c1c-5833-48f8-9ab3-a9a7252915cf">


---

<img width="958" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/13556249/35b1a798-5157-40a6-8bfb-a51d96347d21">

---

<img width="298" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/13556249/63fb290f-1a15-4a26-aa04-ad085736ca0b">

---
```
[INFO] --- vro:2.38.0:run-vro-tests (default-run-vro-tests) @ test2 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[ERROR] /Users/user1/test2/target/vro-tests/test/test2/test2/actions/sample.test.js:10
[ERROR] var sumAction = System.getModule("test2.test2").actionSample;
[ERROR]                                                ^
[ERROR] 
[ERROR] TypeError: Cannot read properties of null (reading 'actionSample')
[ERROR]     at Object.<anonymous> (/Users/user1/test2/target/vro-tests/test/test2/test2/actions/sample.test.js:10:48)
[ERROR]     at Module._compile (node:internal/modules/cjs/loader:1105:14)
[ERROR]     at Module.replacementCompile (/Users/user1/test2/node_modules/@vmware-pscoe/vrotest/node_modules/append-transform/index.js:60:13)
[ERROR]     at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
[ERROR]     at Object.<anonymous> (/Users/user1/test2/node_modules/@vmware-pscoe/vrotest/node_modules/append-transform/index.js:64:4)
[ERROR]     at Module.load (node:internal/modules/cjs/loader:981:32)
[ERROR]     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
[ERROR]     at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:170:29)
[ERROR]     at ModuleJob.run (node:internal/modules/esm/module_job:198:25)
[ERROR]     at async Promise.all (index 0)
[INFO] -----------------|---------|----------|---------|---------|-------------------
[INFO] File             | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
[INFO] -----------------|---------|----------|---------|---------|-------------------
[INFO] All files        |       0 |      100 |       0 |       0 |                   
[INFO]  actionSample.js |       0 |      100 |       0 |       0 | 1-17              
[INFO]  sample.ts       |       0 |      100 |       0 |       0 | 10-14             
[INFO] -----------------|---------|----------|---------|---------|-------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

**Reply by Michaelpalacce** [reply]
<img width="954" alt="image" src="https://github.com/vmware/build-tools-for-vmware-aria/assets/17056014/278961d0-c025-4495-beef-d9f5153204db">

Renaming it to `.js` and removing the type anotations works

**Reply by Michaelpalacce** [reply, marked as answer]
Also you may have a problem since you have the files directly under `src` try. putting them one folder deep. I remember that used to be a problem but I don't think we fixed it

**Comment by unbreakabl3** [comment]
Appreciate. I can confirm the moving of the JS to the sub-folder and adding this folder to the System.getModule() path, solves the issue.
PS. If it's not going to be solved any soon, maybe it worth to update the documentation. I can do it.

**Reply by Michaelpalacce** [reply]
Hey @unbreakabl3 feel free to mention it in the documentation and submit a PR, thank you for that
